src/lib/utils/ct_utils.h

```diff
@@ -75,6 +75,38 @@ constexpr inline void unpoison(T& p) {
    BOTAN_UNUSED(p);
 }
 
+/**
+* This function returns its argument, but (if called in a non-constexpr context)
+* attempts to prevent the compiler from reasoning about the value or the possible
+* range of values. Such optimizations have a way of breaking constant time code.
+*/
+template <typename T>
+constexpr inline T value_barrier(T x) {
+   if(!std::is_constant_evaluated()) {
+      /*
+      * For compilers without inline asm, is there something else we can do?
+      *
+      * For instance we could potentially launder the value through a
+      * `volatile T` or `volatile T*`. This would require some experimentation.
```

Comment:
Looking at [subtle](https://github.com/dalek-cryptography/subtle), they [seem to rely on volatile reads](https://github.com/dalek-cryptography/subtle/blob/6b6a81ad9a6a00c0b42c327eaf4b2f785774377e/src/lib.rs#L217-L243) to prevent the compiler from assuming certain values.

[`core::ptr::read_volatile`](https://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/core/ptr/fn.read_volatile.html) is [translated to](https://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/src/core/ptr.rs.html#453-455) `intrinsics::volatile_load` which, to the best of my understanding, is compiled to a [volatile memory access in LLVM](https://llvm.org/docs/LangRef.html#volatile-memory-accesses) bit code. So perhaps the suggested `volatile` laundering is worth a try.

I'll have a go at it.

Reply:
This does indeed look promising: https://godbolt.org/z/bKTePqYeb

The resulting code is slightly different, but in both cases the compiler optimization is omitted.

Reply:
If you compare the codegen at `-O3` for both GCC and Clang, codegen is much worse when using `volatile`. Many redundant reads/writes to memory

```
        mov     dword ptr [rsp - 4], r8d
        mov     r8d, dword ptr [rsp - 4]
```

Which is fine since we're using inline asm for those compilers anyway.

I'm inclined to merge this as is since the only outlier is MSVC and that compiler is not (that I have heard of) capable of the kind of range analysis that causes these problems with rewriting of constant time code. That might change if MSVC ever switches to LLVM as their backend though!